### PR TITLE
Config file and class

### LINKED ISSUE #9: Add a config file and class
Since this will run as an API, a config file is needed for persistent settings. The config file should be yaml and can mostly be copied over from the DMCI repo.

### PR DESCRIPTION
**Pull request by vkbo**
**Summary**:

Ports over the initialisation of logging and the main Config class from the [dmci](https://github.com/metno/discovery-metadata-catalog-ingestor) repo and adapts the code for the current package. Tests are also copied over.

**Related Issue**:

Closes #9 

**Reviewer checklist**:

* [x] The headers of all files contain a reference to the repository license
* [x] 100% test coverage of new code
* [x] All tests are passing
* [x] All names (e.g., files, classes, functions, variables) are explicit
* [x] Documentation (as docstrings) is complete and understandable

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -11,3 +11,10 @@ This repo is a part of the [Annotated Atlas](https://github.com/metno/AnnotatedA
 ```bash
 pytest -v --cov=ma_search --cov-report=term
 ```
+
+## Logging
+
+The default logging level is `INFO`. To change level, set the environment variable
+`MA_SEARCH_LOGLEVEL` to one of `CRITICAL`, `ERROR`, `WARNiNG`, or `DEBUG`.
```
Comment (fzeiser) [inline README.md:18]:
```suggestion
`MA_SEARCH_LOGLEVEL` to one of `CRITICAL`, `ERROR`, `WARNING`, or `DEBUG`.
```

File: ma_search/__init__.py [hunk 2]
```diff
@@ -17,5 +17,65 @@
...
+    """
+    # Read environment variables
+    wantLevel = os.environ.get("MA_SEARCH_LOGLEVEL", "INFO")
+    logFile = os.environ.get("MA_SEARCH_LOGFILE", None)
+
+    # Determine log level and format
+    if hasattr(logging, wantLevel):
+        logLevel = getattr(logging, wantLevel)
+    else:
+        print("Invalid logging level '%s' in environment variable MA_SEARCH_LOGLEVEL" % wantLevel)
```
Comment (fzeiser) [inline ma_search/__init__.py:40]: shouldn't this be either a warning, or log.warning/log.error?
[not print, which is a sideeffect]

**Comment by vkbo** [comment]
There is no need to review draft pull requests.

File: ma_search/__init__.py (hunk 2 above)
Comment (vkbo) [reply ma_search/__init__.py:40]: This is before logging is fully initialised, so no, it should be print to make sure it is properly returned.

File: ma_search/config.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        self.sqlitePath = None
+
+        # Internals
+        self._rawConf = {}
+
+        return
+
+    def readConfig(self, configFile=None):
+        """Read the config file. If the configFile variable is not set,
+        the class will look for the file in the source root folder.
```
Comment (fzeiser) [inline ma_search/config.py:46]: Add return value

File: ma_search/config.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    ##
+
+    def _readCoreSettings(self):
+        """Read config values under 'main'.
+        """
+        conf = self._rawConf.get("main", {})
+
+        self.dbProvider = conf.get("dbProvider", self.dbProvider)
+
+        return
```
Comment (fzeiser) [inline ma_search/config.py:83]: remove? implicit None return?

File: ma_search/config.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+    def _readSQLiteSettings(self):
+        """Read config values under 'sqlite'.
+        """
+        conf = self._rawConf.get("sqlite", {})
+
+        self.sqlitePath = conf.get("sqlitePath", self.sqlitePath)
+
+        return
```
Comment (fzeiser) [inline ma_search/config.py:92]: remove? implicit None return?
```suggestion
```

File: ma_search/config.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+    def _validateConfig(self):
+        """Check config variable dependencies. It needs to be called after all
```
Comment (fzeiser) [inline ma_search/config.py:95]: Add `Returns`

File: tests/tools.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+# File I/O
+
+def readFile(fileName):
+    """Returns the content of a file as a string.
+    """
```
Comment (fzeiser) [inline tests/tools.py:24]:
``` suggestion
    """ Read contents from file as string

    Returns:
        (str): file content

    Raises:
        OSError: File cannot be read
    """

**Comment by vkbo** [comment]
Please stop reviewing before the PR is complete. It is a waste of your time.

File: ma_search/config.py (hunk 4 above)
Comment (vkbo) [reply ma_search/config.py:83]: Explicit is always better than implicit.

File: ma_search/config.py (hunk 3 above)
Comment (vkbo) [reply ma_search/config.py:46]: This is all copy/pasted code from a different project. I will tell you when the code is ready.

File: ma_search/config.py [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+        # ===============
+        # These should only be set from file.
+
+        # Core Values
+        self.dbProvider = None
+
+        # SQLite Settings
+        self.sqlitePath = None
+
+        return
```
Comment (fzeiser) [inline ma_search/config.py:46]:
```suggestion
```

File: ma_search/__init__.py (hunk 2 above)
Comment (fzeiser) [reply ma_search/__init__.py:45]: still. you may use warnings, which can be filtered away,
```
import warnings
warnings.warn(...)
```

File: ma_search/config.py [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+        Parameters
+        ----------
+        configFile : str or None, optional
+            The config file to be loaded. If None, pgkRoot/config.yaml
+            is attempted loaded instead.
+
+        Returns
+        -------
+        bool
+            True of successful, False if anything failed.
```
Comment (fzeiser) [inline ma_search/config.py:60]:
``` suggestion
True if successful, False if anything failed.
```

**Review by fzeiser — CHANGES_REQUESTED** [review]
There are 2 small comments, otherwise it looks great!

File: ma_search/__init__.py (hunk 2 above)
Comment (vkbo) [reply ma_search/__init__.py:45]: This is an error during initialisation and should absolutely not be filtered away. I don't think this is the right approach. I think in fact this should cause a sys.exit instead.

File: ma_search/config.py (hunk 9 above)
Comment (vkbo) [reply ma_search/config.py:60]: Ok, fixed locally. Will commit when the other one is resolved.

File: ma_search/__init__.py (hunk 2 above)
Comment (fzeiser) [reply ma_search/__init__.py:45]: `raise ValueError`?

Comment (vkbo) [reply ma_search/__init__.py:45]: Maybe. Just need to check what the parent process needs of return codes.

**Review by fzeiser — APPROVED** [review]

**Pull request merged**